cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -1072,6 +1119,57 @@ private ProgramId validateProgramForRunRecord(String namespaceName, String appNa
   }
 
   /**
+   * Helper method to update log levels for Worker, Flow, or Service.
+   */
+  private void updateLogLevels(ProgramId programId, Map<String, LogEntry.Level> logLevels, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
+      if (component == null) {
+        twillController.updateLogLevels(logLevels).get();
+      } else {
+        twillController.updateLogLevels(component, logLevels).get();
+      }
+    }
+  }
+
+  /**
+   * Helper method to reset log levels for Worker, Flow or Service.
+   */
+  private void resetLogLevels(ProgramId programId, @Nullable List<String> loggerNames, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
```

Comment:
The logic of changing a component log level should be encapsulated inside the program controller.